### ISSUE TITLE
Issue in saving image when "iCloud Photo Library" is "ON" in settings in iOS 11

### ISSUE THREAD

**Issue by leftdead7** [terikon/cordova-plugin-photo-library#88]
Getting "Retrieved asset is nil" in failure callback when i am trying to save a image when "iCloud Photo Library" setting is turned ON . When i turn off the setting i am unable to save the image well. 
The plugin was working really well till iOS 10.3 . Please suggest if i need to do any changes for "iCloud Photo Library"

**Comment by leftdead7** [comment]
Sorry

The image is getting saved in the device, but still i am getting "Retrieved asset is nil" in error callback

I actually do not need the item back after saving the image to gallery, so can i safely ignore this error?

**Comment by jayordway** [comment]
I am seeing the same issue with iOS 11.

**Comment by bytestack** [comment]
Same here with iOS 11, disabled iCloud Photo Library and it works know

**Comment by mikewagz** [comment]
Also seeing this.

**Comment by gildebrand** [comment]
I'm having this issue as well, and will do my best to be able to resolve it (even if I'm not very good at native development) 😅 

This actually worked fine while iOS 11.0 was in beta (and maybe even the GM, but I'm not quite sure about that). I will have a look if there was any changes made to the Gallery API during the beta phase.

**Comment by nharrer** [comment]
Could it be a race condition somewhere? Because sometimes it works. 
Also when the exception occurs, the image is not added to the album with the name given in saveImage().
But in either case the image shows up in "All Photos". (Tested on iOS 11.1.1)

**Comment by mubasshir** [comment]
@nharrer did you find any solution?

**Comment by nharrer** [comment]
@mubasshir no, unfortunately not. Thankfully it's not a complete show stopper, since the pictures show up in "All Photos". And I can also confirm that it starts working once iCloud Photo Library is turned off (which is not an option for my customers though).

**Comment by mubasshir** [comment]
I guess I will hardcode and display success message if this particular error occurs

**Comment by endani** [comment]
Hey guys, 
Anyone could solve this? I don't have any more ideas or workarounds here :/

thanks!

**Comment by Saturnyn** [comment]
The issue is discussed here: https://stackoverflow.com/questions/26480526/alassetslibrary-assetforurl-always-returning-nil-for-photos-in-my-photo-stream

The comment by Thomas Hilbert seems to explain the bug in the plugin:

> I've made the observation that trying to retrieve an asset using assetForURL inside a writeImage toSavedPhotosAlbum success block for that same asset will yield an asset of nil (most of the time).
> 
> However, retrieving the asset with assetForURL some time after the writeImage success block has completed execution does yield the correct asset.

I tried his solution and added a `sleep(1)` inside `saveImage()`, just before the `self.putMediaToAlbum()` call. It seems to fix the issue, at least for me.  

This fix is not very clean: I doubt calling sleep() is a good practice, and the 1 second delay will probably be too much or too little depending on the context.  
I don't have enough swift knowledge to write this, but it looks like an acceptable workaround would be to write a loop that calls asset() every 100ms until it returns something else than nil (with a timeout after a few unsuccessful calls).

**Comment by logic17** [comment]
Issue is still present :(

**Comment by zhzhch335** [comment]
![image](https://user-images.githubusercontent.com/13182410/44440743-1b1c9600-a5fc-11e8-8736-6fbb29b845a4.png)
Just sleep it!!! lmao

**Comment by zhzhch335** [comment]
![image](https://user-images.githubusercontent.com/13182410/44440840-86666800-a5fc-11e8-9c92-072698824672.png)
Just sleep it!

**Comment by zhzhch335** [comment]
![image](https://user-images.githubusercontent.com/13182410/44440885-bb72ba80-a5fc-11e8-8ea1-6bda3a05a0f4.png)
Just sleep it!!!
